**Ayşe Zişan Şahin - 22040301132-Yazılım Mühendisliği- Makine Öğrenmesi Vize Ödevi**

In [42]:
# Gerekli kütüphanelerin yüklenmesi
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.decomposition import PCA
from sklearn.discriminant_analysis import LinearDiscriminantAnalysis as LDA
from sklearn.linear_model import LogisticRegression, LinearRegression
from sklearn.tree import DecisionTreeClassifier
from sklearn.naive_bayes import GaussianNB
from sklearn.metrics import confusion_matrix, classification_report, roc_curve, auc,mean_squared_error,accuracy_score

In [5]:
df=pd.read_csv('veriseti.csv')

In [6]:
df.head(5)

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0            6      148             72             35        0  33.6   
1            1       85             66             29        0  26.6   
2            8      183             64              0        0  23.3   
3            1       89             66             23       94  28.1   
4            0      137             40             35      168  43.1   

   DiabetesPedigreeFunction  Age  Outcome  
0                     0.627   50        1  
1                     0.351   31        0  
2                     0.672   32        1  
3                     0.167   21        0  
4                     2.288   33        1

In [7]:
df.isnull().sum()

Pregnancies                 0
Glucose                     0
BloodPressure               0
SkinThickness               0
Insulin                     0
BMI                         0
DiabetesPedigreeFunction    0
Age                         0
Outcome                     0
dtype: int64

-bir DataFrame'in her sütundaki eksik değerlerin sayısını hesaplar.

In [8]:
df.info() #Veri setindeki değerler hakkında bilgi almak için kullanıyoruz.

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 768 entries, 0 to 767
Data columns (total 9 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Pregnancies               768 non-null    int64  
 1   Glucose                   768 non-null    int64  
 2   BloodPressure             768 non-null    int64  
 3   SkinThickness             768 non-null    int64  
 4   Insulin                   768 non-null    int64  
 5   BMI                       768 non-null    float64
 6   DiabetesPedigreeFunction  768 non-null    float64
 7   Age                       768 non-null    int64  
 8   Outcome                   768 non-null    int64  
dtypes: float64(2), int64(7)
memory usage: 54.1 KB


In [9]:
df.eq(0).sum() #Her bir özellikte kaç tane 0 var onu öğrenmek için kullanıyoruz.

Pregnancies                 111
Glucose                       5
BloodPressure                35
SkinThickness               227
Insulin                     374
BMI                          11
DiabetesPedigreeFunction      0
Age                           0
Outcome                     500
dtype: int64

In [10]:
df[['Glucose','BloodPressure','SkinThickness','Insulin','BMI','DiabetesPedigreeFunction','Age']] = df[['Glucose','BloodPressure','SkinThickness','Insulin','BMI','DiabetesPedigreeFunction','Age']].replace(0,np.NaN)

In [11]:
# Eksik değerlerin doldurulması için kullanıyoruz.
df.fillna(df.mean(), inplace=True)

In [12]:
# Normalizasyon
normalized_df = (df - df.min()) / (df.max() - df.min())
print(normalized_df.head(6))

   Pregnancies   Glucose  BloodPressure  SkinThickness   Insulin       BMI  \
0     0.352941  0.670968       0.489796       0.304348  0.170130  0.314928   
1     0.058824  0.264516       0.428571       0.239130  0.170130  0.171779   
2     0.470588  0.896774       0.408163       0.240798  0.170130  0.104294   
3     0.058824  0.290323       0.428571       0.173913  0.096154  0.202454   
4     0.000000  0.600000       0.163265       0.304348  0.185096  0.509202   
5     0.294118  0.464516       0.510204       0.240798  0.170130  0.151329   

   DiabetesPedigreeFunction       Age  Outcome  
0                  0.234415  0.483333      1.0  
1                  0.116567  0.166667      0.0  
2                  0.253629  0.183333      1.0  
3                  0.038002  0.000000      0.0  
4                  0.943638  0.200000      1.0  
5                  0.052519  0.150000      0.0  


-Bu kod, bir DataFrame'deki değerlerin normalizasyonunu gerçekleştirir. Normalizasyon, veri kümesindeki değerleri bir aralığa (genellikle [0, 1] veya [-1, 1] aralığına) dönüştürme işlemidir. Bu, verilerin birbirleriyle karşılaştırılabilir hale gelmesini sağlar ve bazı makine öğrenimi algoritmaları için önemlidir.

In [13]:
# Bağımsız değişkenlerin ve hedef değişkenin belirlenmesi için kullanıyoruz.
X = df.drop("Outcome", axis=1)
y = df["Outcome"]

-Bu kod, tipik bir makine öğrenimi modeli için girdi (X) ve çıktı (y) değişkenlerini ayırmak için kullanılır. Genellikle, makine öğrenimi modelleri bir hedef değişkeni tahmin etmek için bağımsız değişkenleri kullanır. 

In [16]:
sc=StandardScaler()
x1=sc.fit_transform(X)
pca=PCA(n_components=2)
PrincipialComponent=pca.fit_transform(x1)
principialDF=pd.DataFrame(data=PrincipialComponent,columns=["p1","p2"])
principialDF.head()

p1        p2
0  1.503630 -0.588089
1 -1.437513  0.076988
2  0.492233 -0.812569
3 -2.212028  0.118082
4  0.553302  3.469320


-Bu kod, bir veri setindeki özelliklerin (bağımsız değişkenlerin) boyutunu azaltmak için bir dizi dönüşüm işlemi uygular. Genellikle bu tür işlemler, veri görselleştirmesi veya daha az boyutlu veri kümesiyle makine öğrenimi modelleri için çalışma yapmak amacıyla gerçekleştirilir.

In [20]:
# LDA uygulama
lda = LDA(n_components=1)
X_lda = lda.fit_transform(X, y)

-Bu kod, Linear Discriminant Analysis (LDA) yöntemini kullanarak veri kümesinin boyutunu azaltmayı amaçlar. LDA, özelliklerin boyutunu azaltmak ve sınıflar arasındaki farkı maksimize etmek için yaygın olarak kullanılan bir tekniktir

In [22]:
df_lda = pd.DataFrame(X_lda, columns=['LD1'])
print(df_lda.head())

        LD1
0  1.292110
1 -1.765646
2  1.721972
3 -1.733717
4  1.990957


In [25]:
# Veri setinin eğitim ve test setlerine ayrılması için kullanıyoruz.
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

-Bu kod, veri setini eğitim ve test setlerine ayırmak için train_test_split fonksiyonunu kullanır. Bu işlem, genellikle makine öğrenimi modellerini eğitmek ve değerlendirmek için kullanılan bir yöntemdir.

In [30]:
# Çoklu Doğrusal Regresyon analizi için kullanıyoruz.
linear_reg = LinearRegression()
linear_reg.fit(X_train, y_train)
y_pred_linear_reg = linear_reg.predict(X_test)
linear_reg_coefficients = linear_reg.coef_

-Bu kod, çoklu doğrusal regresyon analizi için bir model oluşturur ve bu modeli eğitir. Daha sonra, bu eğitilmiş modeli kullanarak test setinde tahminler yapar ve regresyon katsayılarını elde eder. 

In [31]:
# Multinominal Lojistik Regresyon analizi için kullanıyoruz.
logistic_reg = LogisticRegression(multi_class='multinomial', solver='lbfgs')
logistic_reg.fit(X_train, y_train)
y_pred_logistic_reg = logistic_reg.predict(X_test)

/usr/local/lib/python3.10/dist-packages/sklearn/linear_model/_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(



-Bu kod, multinominal lojistik regresyon analizi için bir model oluşturur ve bu modeli eğitir. 

In [32]:
# Karar Ağacı sınıflandırma algoritması için kullanıyoruz.
decision_tree = DecisionTreeClassifier()
decision_tree.fit(X_train, y_train)
y_pred_decision_tree = decision_tree.predict(X_test)

-Bu kod, karar ağacı sınıflandırma algoritması için bir model oluşturur ve bu modeli eğitir. 

In [33]:
# Naive Bayes sınıflandırıcısı için kullanıyoruz.
naive_bayes = GaussianNB()
naive_bayes.fit(X_train, y_train)
y_pred_naive_bayes = naive_bayes.predict(X_test)


-Bu kod, Naive Bayes sınıflandırıcı modeli için veri setini eğitir ve bu modeli kullanarak test setinde tahminler yapar.

In [34]:
# Performans metriklerinin hesaplanması için kullanıyoruz.
linear_reg_metrics = {
    "MSE": mean_squared_error(y_test, y_pred_linear_reg)
}


-Bu kod, doğrusal regresyon modelinin performansını değerlendirmek için bir metrik olan ortalama karesel hata (MSE) değerini hesaplar.

In [35]:
logistic_reg_metrics = {
    "Accuracy": accuracy_score(y_test, y_pred_logistic_reg),
    "Classification Report": classification_report(y_test, y_pred_logistic_reg),
    "Confusion Matrix": confusion_matrix(y_test, y_pred_logistic_reg)
}


-Bu kod, lojistik regresyon modelinin performansını değerlendirmek için bir dizi metrik hesaplar. 

In [36]:
decision_tree_metrics = {
    "Accuracy": accuracy_score(y_test, y_pred_decision_tree),
    "Classification Report": classification_report(y_test, y_pred_decision_tree),
    "Confusion Matrix": confusion_matrix(y_test, y_pred_decision_tree)
}



-Bu kod, karar ağacı sınıflandırıcının performansını değerlendirmek için bir dizi metrik hesaplar. 

In [37]:
naive_bayes_metrics = {
    "Accuracy": accuracy_score(y_test, y_pred_naive_bayes),
    "Classification Report": classification_report(y_test, y_pred_naive_bayes),
    "Confusion Matrix": confusion_matrix(y_test, y_pred_naive_bayes)
}

-Bu kod, Naive Bayes sınıflandırıcının performansını değerlendirmek için bir dizi metrik hesaplar. 

In [38]:
# Sonuçların raporlanması için kullanıyoruz.
print("Çoklu Doğrusal Regresyon Analizi Sonuçları:")
print("Katsayılar:", linear_reg_coefficients)
print("Mean Squared Error (MSE):", linear_reg_metrics["MSE"])
print()


Çoklu Doğrusal Regresyon Analizi Sonuçları:
Katsayılar: [ 9.58400584e-03  6.56231086e-03 -1.14223519e-03  1.02796804e-04
  2.78844016e-05  1.61661445e-02  4.85119792e-02  5.39767099e-03]
Mean Squared Error (MSE): 0.174317237562553



Bu çıktı, doğrusal regresyon analizinin sonuçlarını içerir. Şimdi sayısal çıktıların yorumunu yapalım:

**Katsayılar (Coefficients)**: Her bir katsayı, bağımsız değişkenlerin (predictors) etkisini temsil eder. Örneğin, "9.58400584e-03" olan bir katsayı, ilgili bağımsız değişkenin bir birim artışının hedef değişkeni 0.00958 birim artıracağını önerir. Pozitif katsayılar, bağımsız değişkenin artmasıyla hedef değişkenin artacağını, negatif katsayılar ise bağımsız değişkenin artmasıyla hedef değişkenin azalacağını gösterir.

**Mean Squared Error (MSE) (Ortalama Karesel Hata)**: MSE, modelin gerçek ve tahmin edilen değerleri arasındaki ortalama karesel farkı gösterir. "0.174317237562553" olan bu değer, modelin tahminlerinin gerçek değerlerden ortalama olarak ne kadar uzak olduğunu belirtir. Düşük MSE değerleri, daha iyi bir model performansını işaret eder.

In [39]:
print("Multinominal Lojistik Regresyon Analizi Sonuçları:")
print("Accuracy:", logistic_reg_metrics["Accuracy"])
print("Classification Report:\n", logistic_reg_metrics["Classification Report"])
print("Confusion Matrix:\n", logistic_reg_metrics["Confusion Matrix"])
print()

Multinominal Lojistik Regresyon Analizi Sonuçları:
Accuracy: 0.7402597402597403
Classification Report:
               precision    recall  f1-score   support

           0       0.79      0.82      0.81       151
           1       0.64      0.59      0.61        80

    accuracy                           0.74       231
   macro avg       0.71      0.70      0.71       231
weighted avg       0.74      0.74      0.74       231

Confusion Matrix:
 [[124  27]
 [ 33  47]]



multinominal lojistik regresyon analizinin sonuçlarını içerir. Şimdi sayısal çıktıların yorumunu yapalım:

**Accuracy (Doğruluk)**: Bu, modelin doğru sınıflandırma oranını ifade eder. "0.7402597402597403" olan bu değer, modelin doğru sınıflandırma oranının %74 olduğunu gösterir. Yani, tüm tahminlerin yaklaşık olarak %74'ünün doğru olduğu anlamına gelir.

**Classification Report (Sınıflandırma Raporu)**: Bu bölüm, her sınıf için precision (kesinlik), recall (duyarlılık) ve f1-score (F1 skoru) gibi metrikleri sunar. Bunlar, modelin sınıflandırma performansını daha ayrıntılı olarak değerlendirmek için kullanılır. Ayrıca, her sınıf için destek (support) değerleri de verilir. Precision, belirli bir sınıfı tahmin ederken ne kadar doğru olduğunu gösterirken, recall bu sınıfın ne kadarını doğru bir şekilde tahmin ettiğini ifade eder. F1-score ise precision ve recall değerlerinin harmonik ortalamasıdır.

**Confusion Matrix (Karmaşıklık Matrisi)**: Bu matris, modelin sınıfları ne kadar doğru ve yanlış tahmin ettiğini gösterir. Soldaki üst köşeden başlayarak, soldan sağa ve yukarıdan aşağıya doğru, gerçek sınıfların ve tahmin edilen sınıfların sayılarını içerir. Örneğin, (0, 0) hücresi, gerçek sınıfın 0 ve tahmin edilen sınıfın da 0 olduğu durumları ifade eder. Bu matriste köşegen dışındaki değerler yanlış sınıflandırmaları gösterir.

In [40]:
print("Karar Ağacı Sınıflandırma Algoritması Sonuçları:")
print("Accuracy:", decision_tree_metrics["Accuracy"])
print("Classification Report:\n", decision_tree_metrics["Classification Report"])
print("Confusion Matrix:\n", decision_tree_metrics["Confusion Matrix"])
print()

Karar Ağacı Sınıflandırma Algoritması Sonuçları:
Accuracy: 0.696969696969697
Classification Report:
               precision    recall  f1-score   support

           0       0.79      0.73      0.76       151
           1       0.55      0.64      0.59        80

    accuracy                           0.70       231
   macro avg       0.67      0.68      0.68       231
weighted avg       0.71      0.70      0.70       231

Confusion Matrix:
 [[110  41]
 [ 29  51]]



karar ağacı sınıflandırma algoritmasının sonuçlarını içerir. Şimdi sayısal çıktıların yorumunu yapalım:

**Accuracy (Doğruluk)**: Bu, modelin doğru sınıflandırma oranını ifade eder. "0.696969696969697" olan bu değer, modelin doğru sınıflandırma oranının %69.7 olduğunu gösterir. Yani, tüm tahminlerin yaklaşık olarak %69.7'sinin doğru olduğu anlamına gelir.

**Classification Report (Sınıflandırma Raporu)**: Bu bölüm, her sınıf için precision (kesinlik), recall (duyarlılık) ve f1-score (F1 skoru) gibi metrikleri sunar. Bunlar, modelin sınıflandırma performansını daha ayrıntılı olarak değerlendirmek için kullanılır. Ayrıca, her sınıf için destek (support) değerleri de verilir. Precision, belirli bir sınıfı tahmin ederken ne kadar doğru olduğunu gösterirken, recall bu sınıfın ne kadarını doğru bir şekilde tahmin ettiğini ifade eder. F1-score ise precision ve recall değerlerinin harmonik ortalamasıdır.

**Confusion Matrix (Karmaşıklık Matrisi)**: Bu matris, modelin sınıfları ne kadar doğru ve yanlış tahmin ettiğini gösterir. Soldaki üst köşeden başlayarak, soldan sağa ve yukarıdan aşağıya doğru, gerçek sınıfların ve tahmin edilen sınıfların sayılarını içerir. Örneğin, (0, 0) hücresi, gerçek sınıfın 0 ve tahmin edilen sınıfın da 0 olduğu durumları ifade eder. Bu matriste köşegen dışındaki değerler yanlış sınıflandırmaları gösterir.

In [41]:
print("Naive Bayes Sınıflandırıcısı Sonuçları:")
print("Accuracy:", naive_bayes_metrics["Accuracy"])
print("Classification Report:\n", naive_bayes_metrics["Classification Report"])
print("Confusion Matrix:\n", naive_bayes_metrics["Confusion Matrix"])

Naive Bayes Sınıflandırıcısı Sonuçları:
Accuracy: 0.7402597402597403
Classification Report:
               precision    recall  f1-score   support

           0       0.81      0.79      0.80       151
           1       0.62      0.65      0.63        80

    accuracy                           0.74       231
   macro avg       0.71      0.72      0.72       231
weighted avg       0.74      0.74      0.74       231

Confusion Matrix:
 [[119  32]
 [ 28  52]]


Naive Bayes sınıflandırıcısının sonuçlarını içerir. Şimdi sayısal çıktıların yorumunu yapalım:

**Accuracy (Doğruluk)**: Bu, modelin doğru sınıflandırma oranını ifade eder. "0.7402597402597403" olan bu değer, modelin doğru sınıflandırma oranının %74 olduğunu gösterir. Yani, tüm tahminlerin yaklaşık olarak %74'ünün doğru olduğu anlamına gelir.

**Classification Report (Sınıflandırma Raporu)**: Bu bölüm, her sınıf için precision (kesinlik), recall (duyarlılık) ve f1-score (F1 skoru) gibi metrikleri sunar. Bunlar, modelin sınıflandırma performansını daha ayrıntılı olarak değerlendirmek için kullanılır. Ayrıca, her sınıf için destek (support) değerleri de verilir. Precision, belirli bir sınıfı tahmin ederken ne kadar doğru olduğunu gösterirken, recall bu sınıfın ne kadarını doğru bir şekilde tahmin ettiğini ifade eder. F1-score ise precision ve recall değerlerinin harmonik ortalamasıdır.

**Confusion Matrix (Karmaşıklık Matrisi)**: Bu matris, modelin sınıfları ne kadar doğru ve yanlış tahmin ettiğini gösterir. Soldaki üst köşeden başlayarak, soldan sağa ve yukarıdan aşağıya doğru, gerçek sınıfların ve tahmin edilen sınıfların sayılarını içerir. Örneğin, (0, 0) hücresi, gerçek sınıfın 0 ve tahmin edilen sınıfın da 0 olduğu durumları ifade eder. Bu matriste köşegen dışındaki değerler yanlış sınıflandırmaları gösterir.